snuba/clusters/cluster.py

```diff
@@ -96,14 +107,39 @@ def get_writer(
 class ClickhouseCluster(Cluster[SqlQuery, ClickhouseWriterOptions]):
     """
     ClickhouseCluster provides a reader, writer and Clickhouse connections that are
-    shared by all storages located on the cluster
+    shared by all storages located on the cluster.
+
+    ClickhouseCluster is initialized with a single address (host/port/http_port),
+    which is used for all read and write operations related to the cluster. This
+    address can refer to either the address of the actual ClickHouse server, or a
+    proxy server (e.g. for load balancing).
+
+    However there are other operations (like some DDL operations) that must be executed
+    on each individual server node, not just on the distributed table. If we are
+    operating a single node cluster, this is straightforward since there is only one
+    server on which to run our command and no distributed table. If we are operating
+    a multi node cluster we need to know the full set of shards and replicas on which
+    to run our commands. This is provided by the `get_nodes()` method.
     """
 
-    def __init__(self, host: str, port: int, http_port: int, storage_sets: Set[str]):
+    def __init__(
+        self,
+        host: str,
+        port: int,
+        http_port: int,
+        storage_sets: Set[str],
+        single_node: bool,
+        # The cluster name if single_node is set to False
+        cluster_name: Optional[str] = None,
```

Comment:
We started the other long thread about the single_node issue and then it turned into a different topic (query vs storage nodes). Let's pick up the conversation on single_node here.

I don't think we need to pass the single node flag and store it for now (unless you plan to use it right away for the table name) it is not needed to return the proper list of nodes.

Reply:
I'm not sure I follow why it's not needed. Is is used to return the proper list of nodes - that either needs to come from the provided host parameter if single_node = true or `system.clusters` if single_node = false. It would also be used for the table name and anything else that currently depends on the global `DATASET_MODE` setting which would become deprecated in favor of cluster specific values.

Reply:
If you prepopulate the list of hosts at initialization you do not need to store it.
About the table name, my only concern is that I still have not clear who is supposed to do the mapping from an abstract name to the actual name on clickhouse. Is this going to be the responsibility of the Cluster object ?

Reply:
Right now table name is resolved by the table schema. As far as I'm aware it cannot be done by the storage since one storage may be backed by more than one physical table. I am not sure whether either 1) the schema should become aware of whether it's in the context of single node cluster or multi node cluster, or 2) this should be done somewhere else entirely.